src/storage/src/table/cell_based_table.rs

```diff
@@ -232,83 +301,94 @@ impl<S: StateStore> CellBasedTable<S> {
             deserializer.deserialize(&key, &value).map_err(err)?;
         }
 
-        let pk_and_row = deserializer.take();
-        Ok(pk_and_row.map(|(_pk, row)| row))
+        let result = deserializer.take();
+        Ok(result.map(|(vnode, _pk, row)| {
+            // This table should only to used to access entries with vnode specified in
+            // `self.vnodes`.
+            assert!(self.vnodes.is_set(vnode as usize).unwrap());
+            row
+        }))
+    }
+}
+
+/// Write
+impl<S: StateStore> CellBasedTable<S, READ_WRITE> {
+    /// Get vnode value with full row.
+    fn compute_vnode_by_row(&self, row: &Row) -> VirtualNode {
+        // With `READ_WRITE`, the output columns should be exactly same with the table columns, so
+        // we can directly index into the row with indices to the table columns.
+        let vnode = match self.dist_key_indices.as_ref() {
+            Some(indices) => row
+                .hash_by_indices(indices, &CRC32FastBuilder {})
+                .unwrap()
+                .to_vnode(),
+            None => DEFAULT_VNODE,
+        };
+        // This table should only to used to access entries with vnode specified in `self.vnodes`.
+        assert!(self.vnodes.is_set(vnode as usize).unwrap());
+        vnode
     }
 
-    async fn batch_write_rows_inner<const WITH_VALUE_META: bool>(
+    /// Write to state store.
+    pub async fn batch_write_rows(
         &mut self,
         buffer: BTreeMap<Vec<u8>, RowOp>,
         epoch: u64,
     ) -> StorageResult<()> {
-        debug_assert!(self.is_complete(), "cannot write to a partial table");
-
-        // stateful executors need to compute vnode.
         let mut batch = self.keyspace.state_store().start_write_batch();
         let mut local = batch.prefixify(&self.keyspace);
 
         for (pk, row_op) in buffer {
-            // If value meta is computed here, then the cell based table is guaranteed to have
-            // distribution keys. Also, it is guaranteed that distribution key indices will
-            // not exceed the length of pk. So we simply do unwrap here.
             match row_op {
                 RowOp::Insert(row) => {
-                    let value_meta = if WITH_VALUE_META {
-                        ValueMeta::with_vnode(self.compute_vnode_by_row(&row))
-                    } else {
-                        ValueMeta::default()
-                    };
+                    let vnode = self.compute_vnode_by_row(&row);
```

Comment:
Should vnode be computed by cell based table or state table? If we are to pass vnode from upper layer and do filtering accordingly, then memtable should also have the info of vnode.

Reply:
What do you mean by "do filtering accordingly"? 🤔 Currently the memtable associated with the cell-based table will also be partitioned into the same vnode, so there's no need to do filtering.

Reply:
For example, executor only need the data of a specific vnode. (Is that possible?)

Reply:
The streaming executor should tell the cell-based table the vnodes it cares about when constructing a `CellBasedTable` instance. The vnode bitmap changes only if the system scales in/out, and the executors will be dropped and rebuilt according to our design.

The batch executors may need this, which is unrelated to the memtable, however. Will implement in the following PRs.